### PR TITLE
[tree,io] prevent integer overflow when dealing with cachesize

### PR DESCRIPTION
**Pull request by ferdymercury**
# This Pull request:

## Changes or fixes:

Fixes https://github.com/root-project/root/issues/9292

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

### CODE REVIEW

**Comment by ferdymercury** [comment]
@pcanal is there a way to run the CI against this roottest PR ? https://github.com/root-project/roottest/pull/1069/files

File: tree/tree/src/TTreeCloner.cxx [hunk 1]
```diff
@@ -644,12 +644,13 @@ void TTreeCloner::ImportClusterRanges()
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// Set the TFile cache size to be used.
+/// Set the cache size used by the matching TFile.
 /// Note that the default is to use the same size as the default TTreeCache for
 /// the input tree.
-/// \param size Size of the cache. Zero disable the use of the cache.
+/// \param size Size of the cache. Zero (or a value not matching the buffer size
```
Comment (pcanal) [inline tree/tree/src/TTreeCloner.cxx:650]: This change ("or a value not matching ...) is inaccurate.  What the code does is delete the existing file cache if it is not matching the request size (or disabling) so that (as pointed to by the comment at the end) the file cache can be created later with the right size (or not create if the size is 0).

**Comment by pcanal** [comment]
> is there a way to re-run the CI against this roottest PR ? [root-project/roottest#1069 (files)](https://github.com/root-project/roottest/pull/1069/files)

If/when the new CI or jenkins build are re-run they will pick up the new branch since it has the same name as this PR's branch.

However the change is surprising:
```
-Estimate cluster size: 3
+Estimate cluster size: 3030
```
does not seem like it is close to the limits and is unclear why it is changed.

File: io/io/src/TFileCacheRead.cxx [hunk 2]
```diff
@@ -711,7 +711,7 @@ void TFileCacheRead::WaitFinishPrefetch()
 Int_t TFileCacheRead::SetBufferSize(Long64_t buffersize)
 {
    if (buffersize <= 0) return -1;
-   if (buffersize < 10000) buffersize = 100000;
+   if (buffersize <=10000) buffersize = 100000;
```
Comment (pcanal) [inline io/io/src/TFileCacheRead.cxx:714]: Why this change?

**Comment by ferdymercury** [comment]
> > is there a way to re-run the CI against this roottest PR ? [root-project/roottest#1069 (files)](https://github.com/root-project/roottest/pull/1069/files)
> 
> If/when the new CI or jenkins build are re-run they will pick up the new branch since it has the same name as this PR's branch.
> 
> However the change is surprising:
> 
> ```
> -Estimate cluster size: 3
> +Estimate cluster size: 3030
> ```
> 
> does not seem like it is close to the limits and is unclear why it is changed.

This is because, there was some clamping happening, but the value of cacheSize was not being taken into account, as this line was missing in the code:
`cacheSize = pf->GetBufferSize();`

File: tree/tree/src/TTreeCloner.cxx (hunk 1 above)
Comment (ferdymercury) [reply tree/tree/src/TTreeCloner.cxx]: fixed

File: tree/tree/src/TTree.cxx [hunk 3]
```diff
@@ -8778,6 +8782,7 @@ Int_t TTree::SetCacheSizeAux(bool autocache /* = true */, Long64_t cacheSize /*
          if (res < 0) {
             return -1;
          }
+         cacheSize = pf->GetBufferSize(); // update after potential clamp
```
Comment (ferdymercury) [inline tree/tree/src/TTree.cxx:8785]: @pcanal this is the line I was referring to. a cacheSize is requested which then has nothing to do with the actual cacheSize set because of the inherent clamping, thus an update of the value is needed for consistency, otherwise one gets unrealistic results. That's why the roottest had to be adapted.

File: io/io/src/TFileCacheRead.cxx [hunk 4]
```diff
@@ -701,15 +702,17 @@ void TFileCacheRead::WaitFinishPrefetch()
...
 /// Return values:
 ///   - 0 if the prefetch block lists remain unchanged
 ///   - 1 if some or all blocks have been removed from the prefetch list
 ///   - -1 on error
 
-Int_t TFileCacheRead::SetBufferSize(Int_t buffersize)
+Int_t TFileCacheRead::SetBufferSize(Long64_t buffersize)
 {
    if (buffersize <= 0) return -1;
    if (buffersize <=10000) buffersize = 100000;
```
Comment (ferdymercury) [inline io/io/src/TFileCacheRead.cxx:714]: @pcanal Is this clamping intended, or is it a typo? (10k vs 100k) 

Otherwise, I rebased and looks ready to me. Thanks for the review!

Comment (pcanal) [reply io/io/src/TFileCacheRead.cxx:714]: I think it was intentional, it the cache is too small (10k) ignore the rest and set it to something 'reasonable'

**Review by pcanal — APPROVED** [review]
LGTM.  Thanks.

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
@pcanal, @ferdymercury, this PR caused a regression in a unit test that I see in other PRs, e.g.:
https://github.com/root-project/root/pull/15173

**Comment by pcanal** [comment]
Yes, the companion roottest PR was merge a couple of hours ago.
